docs/sphinx/specification/cudaq/types.rst

```diff
@@ -4,30 +4,91 @@ The CUDA Quantum language extension provides certain fundamental types that are
-this system (for :math:`N` qudits) can be described by a :math:`D`\ :sup:`N`\-dimensional vector in
-Hilbert space with the absolute square of all elements summing to 1. 
+Quantum dit (qudit)
+===================
 
-**[2]** The :code:`cudaq::qudit` encapsulates a unique :code:`std::size_t` modeling the index of the
-qudit in the underlying quantum memory space (assuming an infinite register
-of available qudits). 
+The quantum dit, or qudit for short, is a model to represent a :math:`d`-level 
+(or :math:`d`-dimension) quantum system. Mathematically, we describe the state
+of a qudit using a :math:`d`-dimensional vector of unit form, living in the
+vector space :math:`\mathbb{C}^d` (a Hilbert space). Qudits with known values
+for :math:`d` have specific name. A qubit has dimension 2, a qutrit has
+dimension 3, and so on.
 
-**[4]** To adhere to the no-cloning theorem of quantum mechanics,
-the :code:`cudaq::qudit` is non-copyable and non-movable. Therefore, all :code:`cudaq::qudit` 
-instances must be passed by reference, and the no-cloning theorem is satisfied
-at compile-time. 
+In a CUDA Quantum implementation, objects of type qudit must adhere to the
+following constraints:
 
-**[5]** :code:`cudaq::qudit` instances can only be allocated within CUDA Quantum quantum
-kernel code and can never be allocated from classical host code.
+**[1]** Objects of type :code:`qudit` shall be parameterized with :math:`d`,
+        to indicate it's dimension.
 
-**[6]** All instantiated :code:`cudaq::qudit` instances start in the :code:`|0>` computational basis 
-state and serve as the primary input argument for quantum intrinsic operations 
-modeling typical logical quantum gate operations.
+**[2]** CUDA Quantum kernels shall not have objects of type :code:`qudit` with
+        different dimensions.
 
-**[7]** :code:`cudaq::qudit` instances can be negated when leveraged as controls in 
-quantum operation invocations. The mechanism for negation should be via overloading of :code:`qudit<N>::operator!()`.
+**[3]** Objects of type :code:`qudit` shall not carry a quantum state.
+
+**[4]** Objects of type :code:`qudit` shall identify a qudit in the underlying
+        quantum system. The identifier shall be unique and valid during the
+        lifetime of the :code:`qudit` object.
+
+**[5]** Objects of type :code:`qudit` and any directly or indirectly derived
+        type shall only be used within quantum kernel code.
+
+**[6]** The default constructor shall create an object of type code:`qudit` that
+        references a qudit in the :math:`|0\rangle` computational basis state.
+
+**[7]** Objects of type :code:`qudit` may be non-copyable.
+
+**[8]** Objects of type :code:`qudit` may be non-movable.
+
+**[9]** Objects of type :code:`qudit` may have a mechanism that indicates the
+        presence of implicit Pauli-X operators surrounding the qudit use as a
+        control. For example (pseudocode):
+        .. code-block:: cpp
```

Comment:
with these types of snippets I've tried to do the C++ and Python tabs.

Reply:
Also, I think you may need a newline between the paragraph and the code block (may not render correctly without)